### PR TITLE
Websockets Support

### PR DESCRIPTION
**Pull request by benjamincburns**
This PR supersedes #14, as it is based on those changes. Thanks heaps to @perissology for his submission!

How this all works:

Subscription creation and notification:

- In MetaMask/provider-engine#207, I created a `SubscriptionSubprovider` which extends the existing `FilterSubprovider`
- When the `SubscriptionSubprovider` gets an `eth_subscribe` request, it uses the underlying `FilterSubprovider` code to create a Filter (i.e. `LogFilter`, `BlockFilter`, or `PendingBlockFilter`) for the subscription, then returns the id of this filter.
- I refactored the `LogFilter`, `BlockFilter` and `PendingBlockFilter` types to emit a `data` event on update
- When `SubscriptionSubprovider` creates a filter, it subscribes to the filter's `data` event, and and emits its own `data` event when the Filter's event fires, clearing pending filter changes in the process.
- Similarly, I refactored our Provider type to be an `EventEmitter` and to proxy `data` events from the `SubscriptionSubprovider`.

Routing notifications to clients:

- As one might expect, we listen on a WebSocket, and proxy RPC calls over to the existing `Provider` type.
- The component responsible for this proxying is the `ConnectionManager`.
- The `ConnectionManager` snoops on the traffic it's proxying to maintain a forward and reverse mapping of subscription ID to websocket connection.
- The `ConnectionManager` listens to `data` events from the Provider, and uses its mapping of subscription ID to connection to route the subscription notifications to the correct websocket connection.
- When an `eth_unsubscribe` request is handled, `ConnectionManager` cleans up the relevant mapping.
- When a client disconnects, `ConnectionManager` cleans up all mappings pertaining to that client.

TODO: 
- [x] Fix intermittent forking test failure
- [x] Fix block tracker to avoid polling
- [x] Rebase on current develop
- [x] Squash a bunch of noisy commits (especially the one where I misspelled "concurrency")

### CODE REVIEW

**Review by gnidan — COMMENTED** [review]
Hey @benjamincburns - couple comments / thoughts from a first read.

Looks great! Lots of bits and pieces moving around, but generally seems pretty intuitive.

Hat tip 👒 to @perissology for getting this started!

File: lib/blockchain_double.js [hunk 1]
```diff
@@ -219,7 +225,12 @@ BlockchainDouble.prototype.putBlock = function(block, logs, receipts, callback)
       );
     });
 
-    async.parallel(requests, callback);
+    async.parallel(requests, (err, result) => {
+      if (!err) {
+        self.emit('block', block);
+      }
+      callback(err, result);
+    });
```
[the remark this answers is not in the document]
Comment (benjamincburns) [reply lib/blockchain_double.js]: There shouldn't be. `pushBlock` is only called when `processBlock` wants to commit a new block to the chain, when the genesis block is created, or on start with an existing db when `latest` is read from that db. None of those things should be possible to order incorrectly.

File: lib/server.js [hunk 2]
```diff
@@ -119,11 +55,16 @@ module.exports = {
 
     server.provider = provider;
 
+    var wss = null;
+
+    if (options.ws) {
+      wss = webSocketServer(server, provider, logger);
+    }
```
[the remark this answers is not in the document]
Comment (benjamincburns) [reply lib/server.js]: I thought so at first as well,  but after testing, I think this is the correct way to do it. `webSocketServer` has a reference to the `httpServer`, meaning it listens on the same socket, and its connections are tracked/managed by the `httpServer` anyway.

File: lib/statemanager.js [hunk 3]
```diff
@@ -319,12 +319,6 @@ StateManager.prototype.queueTransaction = function(method, tx_params, callback)
   // Get the nonce for this address, taking account any transactions already queued.
```
[the remark this answers is not in the document]
Comment (benjamincburns) [reply lib/statemanager.js]: good catch

File: lib/statemanager.js [hunk 4]
```diff
@@ -342,13 +336,12 @@ StateManager.prototype.queueTransaction = function(method, tx_params, callback)
     self.action_queue.push({
       method: method,
       from: tx_params.from,
-      tx: tx,
+      tx: rawTx,
       callback: callback
     });
 
     // We know there's work, so get started.
     self.processNextAction();
```
[the remark this answers is not in the document]
Comment (benjamincburns) [reply lib/statemanager.js]: good catch

File: lib/statemanager.js [hunk 5]
```diff
@@ -533,37 +535,60 @@ StateManager.prototype.processCall = function(from, tx, callback) {
...
-  self.blockchain.queueTransaction(tx);
+  self.blockchain.getQueuedNonce(from, (err, nonce) => {
+    // If the user specified a nonce, use that instead.
+    if (rawTx.nonce == null) {
+      // account for transactions waiting in the tx queue
+      rawTx.nonce = to.hex(nonce);
+    }
 
-  var tx_hash = to.hex(tx.hash());
+    let tx = new FakeTransaction(rawTx);
```
[the remark this answers is not in the document]
Comment (benjamincburns) [reply lib/statemanager.js]: definitely.

File: lib/webSocketServer.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+
+ConnectionManager.prototype.manageConnection = function (connection) {
+  const self = this;
+  connection.id = ++self.connectionCounter;
+  self.connections[ connection.id ] = {
+    connection: connection,
+    subscriptions: [],
+    subscriptionCounter: 0,
```
[the remark this answers is not in the document]
Comment (benjamincburns) [reply lib/webSocketServer.js]: I'll double check, I think you're correct. I think it's likely hold over from the initial set of refactored changes.

[the remark this answers is not in the document]
Comment (benjamincburns) [reply lib/webSocketServer.js]: I didn't get to this yesterday. Will fix this now.

File: test/runtime_errors.js [hunk 7]
```diff
@@ -76,4 +82,10 @@ describe("Runtime Errors", function() {
     });
   });
 
+  after('shutdown', function(done) {
+    let provider = web3._provider;
+    web3.setProvider();
+    provider.close(done);
```
Comment (tcoulter) [inline test/runtime_errors.js]: What does `web3.setProvider()` do when not passing in a provider as a parameter?

Comment (benjamincburns) [reply test/runtime_errors.js]: from memory it sets the provider to `undefined` (it doesn't check the arg). I think I was being overly pedantic here though. My thought process was that web3 has a bunch of stuff it does in the background, and I didn't want it trying to run a request during or after provider close.

Further, I think calling `provider.close` is a good idea in general when using the direct provider (probably need to make sure other tests do this), but unfortunately it's not a standard part of the interface, so I need to resort to something like `provider.connection.close` for the WebSocket provider. Maybe I'll raise an issue for that.

**Pull request merged**